src/java.base/share/classes/java/util/concurrent/ForkJoinPool.java

```diff
@@ -2094,63 +2097,59 @@ else if (o == null)       // contended
-                            if (((int)c & SMASK) == (idlePhase & SMASK) &&
-                                compareAndSetCtl(c, nc)) {
-                                w.source = DEREGISTERED;
-                                w.phase = (int)c;
-                                break;
-                            }
-                            deadline += keepAlive;   // not head; reset timer
-                        }
-                    }
-                    w.parker = null;
-                    LockSupport.setCurrentBlocker(null);
+        if ((p & IDLE) != 0 || w == null)       // already terminating
+            return p;
+        int nextPhase = p + (IDLE << 1);
+        long pc = ctl, qc = (nextPhase & LMASK) | ((pc - RC_UNIT) & UMASK);
+        w.stackPred = (int)pc;                  // set ctl stack link
+        w.phase = p | IDLE;                     // try to inactivate
+        if (!compareAndSetCtl(pc, qc))          // contended enque
+            return w.phase = p;                 // back out
+        WorkQueue[] qs = queues;                // recheck queues
+        int n = (qs == null) ? 0 : qs.length;
+        for (int l = n, r = nextPhase; l > 0; --l, ++r) {
+            WorkQueue q; ForkJoinTask<?>[] a; int cap;
+            if ((q = qs[r & SMASK & (n - 1)]) != null && (a = q.array) != null &&
+                (cap = a.length) > 0 && a[q.base & (cap - 1)] != null &&
+                qc == ctl && qc == compareAndExchangeCtl(qc, pc))
+                return p = w.phase = nextPhase; // self-signal
+            if ((p = w.phase) == nextPhase)
+                return p;                       // interleave signal checks
+            Thread.onSpinWait();                // reduce memory traffic
+        }
+        long deadline = 0L;                     // timeout if trimmable
+        if ((qc & RC_MASK) <= 0L && quiescent() && // all idle and w is ctl top
+            (nextPhase & SMASK) == ((int)(qc = ctl) & SMASK)) {
+            int nt = (short)(qc >>> TC_SHIFT), np = parallelism;
+            long delay = keepAlive;
+            if (np != nt)                       // scale if not at target
+                delay = Math.max(TIMEOUT_SLOP, delay / Math.max(2, np));
+            if ((deadline = delay + System.currentTimeMillis()) == 0L)
+                deadline = 1L;                  // avoid zero
+        }
+        LockSupport.setCurrentBlocker(this);    // emulate LockSupport.park
+        w.parker = Thread.currentThread();
+        for (p = IDLE;;) {
+            if ((runState & STOP) != 0 || (p = w.phase) == nextPhase)
+                break;
+            U.park(deadline != 0L, deadline);
+            if ((p = w.phase) == nextPhase || (runState & STOP) != 0)
+                break;
+            Thread.interrupted();               // clear for next park
+            if (deadline != 0L &&               // try to trim on timeout
+                deadline - System.currentTimeMillis() < TIMEOUT_SLOP) {
```

Comment:
Since `currentTimeMillis()` is fetched from system time, I think it might be more appropriate to use nanoTime() to track delays.

Reply:
It would be better, but absolute-time Unsafe.park only operates at millisecond accuracy (and even at that may misfire early, requiring TIMEOUT_SLOP.)

Reply:
Yes. But because this is a performance not correctness issue, it's OK to (rarely) speed up or slow down trims.